### DISCUSSION TITLE
Перенос треков из чартов Shazam в Spotify

### DISCUSSION THREAD

**Discussion by Maracumba** [Chimildic/goofy#161] [Идея] [answered]
Есть такая хотелка... Оставлю здесь, получится или нет.
Создать функцию по получению списков треков из Shazam:
- Discovery 
- Топы по жанрам

**Comment by Chimildic** [comment]
Spotify поиск среди плейлистов по слову `shazam` не дает нужных результатов?

**Reply by Maracumba** [reply]
Там есть множество плейлистов, более похожих на мусор)).
В принципе функция в гуфи не обязательна.
Может можно просто как то сделать так ручками?
-взять со спотифай csv файл со списком треков
-преобразовать его ручками под формат файла кэша для гуфи
-закинуть на гугл диск
-потом в гуфи через этот файл искать через поиск Spotify, по типу как с фм радио сделано

Можно так сделать?

**Comment by Chimildic** [comment, marked as answer]
Добавьте функцию `shazamToSpotify_` в проект (в новый файл или существующий для аддонов). 
```js
function shazamToSpotify_(path) {
  let url = `https://www.shazam.com/services/charts/csv/${path}`;
  let text = CustomUrlFetchApp.fetch(url).getContentText();
  let csvStr = text.split('\n').slice(2);
  let keys = csvStr.splice(0, 1)[0].split(',');
  let shazamTracks = csvStr.reduce((array, str) => {
    let items = str.split(',');
    let obj = keys.reduce((obj, key, i) => {
      obj[key.toLowerCase()] = items[i].formatName().replace(/"/g, '');
      return obj;
    }, {});
    array.push(obj);
    return array;
  }, []);
  return Search.multisearchTracks(shazamTracks, (item) => `${item.artist} ${item.title}`);
}
```

Для получения Spotify треков укажите путь до необходимой страницы после слова `charts`. Например, если ссылка `https://www.shazam.com/ru/charts/genre/russia/pop`, то путь для функции `genre/russia/pop`.

```js
let topWorldTracks = shazamToSpotify_('top-200/world');
let discoveryAustriaTracks= shazamToSpotify_('discovery/austria');
let danceRussiaTracks = shazamToSpotify_('genre/russia/dance');
```
> Поиск каждого трека тратит один запрос

**Reply by Maracumba** [reply]
Как всегда на высоте
